### PR TITLE
Update Google Libraries

### LINKED ISSUE #1186: Update Google Ads library
To be able to use the new [Performance Max Campaigns](https://developers.google.com/google-ads/api/docs/performance-max/create-campaign) we will need to update to at least [v9 of the Google Ads API](https://developers.google.com/google-ads/api/docs/release-notes#v9)

The earliest version that supports v9 is the PHP library version [12.1.0](https://github.com/googleads/google-ads-php/releases/tag/v12.1.0) (currently using 10.0.0).

Note: Version [13.0.0](https://github.com/googleads/google-ads-php/releases/tag/v13.0.0) is also available but that requires the minimum PHP version to be bumped to 7.4 (we still have the minimum as 7.3).

Points to keep an eye on:
- google/protobuf is updated to 3.18.0 this package is shared with dependencies of the Merchant Center libraries
- Confirm all namespace replacements in bin/prefix-vendor-namespace.php still work and we aren't missing any new ones for updated packages (or can remove old replacements no longer needed)
- Remove autoload file for `guzzlehttp/psr7`, see last comment about needing to update Google API library for this
- WCS can be updated to v9 along with the plugin update (see example PR 1807-gh-Automattic/woocommerce-connect-server )
- Confirm we don't get any MISALIGNED_DATE_FOR_FILTER errors when using various date filters for the Ads reports, see [migration details](https://developers.google.com/google-ads/api/docs/version-migration#v8-v9)

### LINKED ISSUE #1199: Update google API library to 2.12.1
After updating the Ads library to v10 we no longer required the fallback functions, however there were still some dependencies in `google/gax` and `google/auth` that required `guzzlehttp/psr7 ^1.2`. These have now all been changed to allow `^2.0`, so we no longer need to autoload the [fallback file](https://github.com/woocommerce/google-listings-and-ads/blob/1.10.0/composer.json#L40), for Google Ads, see #957.

However the Google API library also has a dependency on the `google/auth` package. So we would need to update to version [2.12.1](https://github.com/googleapis/google-api-php-client/releases) to fully remove the autoloaded file.

After updating we also need to check the full dependencies for `guzzlehttp/guzzle` and `guzzlehttp/promises` to confirm if we can remove those autoloads as well.

### PR DESCRIPTION
**Pull request by mikkamp**
### Changes proposed in this Pull Request:

Since both the content API library and the ads API library share common packages, I updated both of them in addition to all their dependencies.

Besides updating the package we have some version dependent code for the Ads API, which needed to be bumped from version V8 > V9. To connect with the Ads API we maintain a custom copy of the class GoogleAdsClient and the trait ServiceClientFactoryTrait. These are both version dependent and I updated them to include the new services.

The package `guzzlehttp/psr7` no longer has an autoload for a `functions_include.php` file, so this was removed from composer.json. Both `guzzlehttp/guzzle` and `guzzlehttp/promises` still have a file so this was left in place.

The code to directly access the function `GuzzleHttp\Psr7\str` is no longer present in the replaced packages, so it was removed from the list of functions where we replace the namespace. The other ones in the list remain in place, I did a search for new functions but this came up empty.

In the composer.json we run a script which removes older versions of the Ads Library, since we switched to version 9, I added version 8 to the list of versions to cleanup.

A pending task would be to update the WCS to also use V9 for any of the manager requests.

Closes #1186 
Closes #1199 

### Overview of updated packages

`composer update googleads/google-ads-php --with-all-dependencies`

  - Upgrading firebase/php-jwt (v5.4.0 => v5.5.1)
  - Upgrading google/auth (v1.16.0 => v1.18.0)
  - Upgrading google/common-protos (1.3 => 1.4.0)
  - Upgrading google/gax (v1.7.1 => v1.10.0)
  - Upgrading google/grpc-gcp (0.1.5 => v0.2.0)
  - Upgrading google/protobuf (v3.17.3 => v3.19.3)
  - Upgrading googleads/google-ads-php (v10.0.0 => v12.1.0)
  - Upgrading grpc/grpc (1.38.0 => 1.42.0)
  - Upgrading guzzlehttp/guzzle (7.3.0 => 7.4.1)
  - Upgrading guzzlehttp/promises (1.4.1 => 1.5.1)
  - Upgrading guzzlehttp/psr7 (1.8.2 => 2.1.0)
  - Upgrading monolog/monolog (2.2.0 => 2.3.5)
  - Locking psr/http-factory (1.0.1)
  - Upgrading symfony/deprecation-contracts (v2.2.0 => v2.5.0)
  
`composer update google/apiclient --with-all-dependencies`

  - Upgrading google/apiclient (v2.11.0 => v2.12.1)
  - Upgrading google/apiclient-services (v0.216.0 => v0.230.0)
  - Upgrading paragonie/constant_time_encoding (v2.4.0 => v2.5.0)
  - Upgrading phpseclib/phpseclib (3.0.9 => 3.0.12)

### Detailed test instructions:
1. Check out this branch and remove any previous vendor folder
2. Run composer install
3. Check package versions `composer show --no-dev`
4. Confirm the google packages with the following versions
```
google/apiclient                    v2.12.1   Client library for Google APIs
google/apiclient-services           v0.230.0  Client library for Google APIs
google/auth                         v1.18.0   Google Auth Library for PHP
google/common-protos                1.4.0     Google API Common Protos for PHP
google/gax                          v1.10.0   Google API Core for PHP
google/grpc-gcp                     v0.2.0    gRPC GCP library for channel management
google/protobuf                     v3.19.3   proto library for PHP
googleads/google-ads-php            v12.1.0   Google Ads API client for PHP
grpc/grpc                           1.42.0    gRPC library for PHP
guzzlehttp/guzzle                   7.4.1     Guzzle is a PHP HTTP client library
guzzlehttp/promises                 1.5.1     Guzzle promises library
guzzlehttp/psr7                     2.1.0     PSR-7 message implementation that also provides common utility methods
```
5. Test the plugin and the Connection Test page and confirm we can send requests without any fatals or errors (need to connect all accounts)
6. Send both regular requests to Merchant Center and Ads as well as proxied requests in the Connection Test page.

### Testing for the Ads reports
To confirm we don't get any `MISALIGNED_DATE_FOR_FILTER` errors, we need to test requests for report data about campaigns. This will need to be done either with an account that has live campaigns or we can use a local copy of the WCS modified to show live report data.

1. Setup a local copy of a modified WCS (see p1642505834083200-slack-C01E9LB4X61 )
2. Go to Marketing > Google Listings & Ads > Reports > Products (select from Paid Campaigns) and confirm we see report data
3. Select various date ranges to confirm we still get data
4. The reports pages only support an interval of `day` at the moment. So to really test the error we will need to confirm that different intervals work, which can be done with a manual request such as through Postman
5. Send a GET request through Postman with an interval set to weekly:
```
https://domainname.test/wp-json/wc/gla/ads/reports/products?after=2022-01-04&before=2022-01-15&interval=week&order=desc&orderby=spend&fields[0]=sales&fields[1]=spend&fields[2]=clicks&fields[3]=impressions&fields[4]=conversions
```
6. Repeat the request with several different date ranges making sure the after doesn't start the same time as the start of the week. Other interval values can be tested as well, such as `quarter`, `month` or `year`

### Changelog entry
* Update - Google Ads API to V9.
* Update - Google Content API library to 2.12.1.

### CODE REVIEW

File: src/Google/Ads/ServiceClientFactoryTrait.php [hunk 1]
```diff
@@ -170,6 +190,8 @@ public function getGoogleAdsClientOptions(): array {
 			self::$CREDENTIALS_LOADER_KEY => $this->getOAuth2Credential(),
 			self::$DEVELOPER_TOKEN_KEY    => '',
 			self::$TRANSPORT_KEY          => 'rest',
+			'libName'                     => Constants::LIBRARY_NAME,
+			'libVersion'                  => Constants::LIBRARY_VERSION,
```
Comment (puntope) [inline src/Google/Ads/ServiceClientFactoryTrait.php:194]: 💅  WDYT about saving those 2 as keys as we did for `self::$TRANSPORT_KEY`?

**Review by puntope — COMMENTED** [review]
The code changes make sense for me ✅ 
The Connection test cheks were performed well ✅ 
The composer libs were updated as expected ✅ 

❓ I have doubts about the Ads Reports part... I was able to check last year reports, but not the new ones with the provided account account.

❓  The Get request you provided causes error on my Postman

Screenshot 2022-01-28 at 15 15 03

**Comment by puntope** [comment]
> ❓ I have doubts about the Ads Reports part... I was able to check last year reports, but not the new ones with the provided id account.

The steps I followed were to apply the patch, connect WP.com, Google and Merchant... and after that change the ID in the line 33 of `lib/google-proxy/index.js` and changed the ids...  Checked also in incognito

File: src/Google/Ads/ServiceClientFactoryTrait.php (hunk 1 above)
Comment (mikkamp) [reply src/Google/Ads/ServiceClientFactoryTrait.php:194]: I think it's a good idea. Although this code is a modified version of the Trait which is included in the library, so I was trying to stick as close to the original as possible, so we have less maintenance when updating this file. So I'd prefer to leave it as is.

**Review by puntope — APPROVED** [review]
✅  LGTM

After applying the correct account config I can see the reports and query the reports as expected.

**Pull request merged**